src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs

```diff
@@ -2060,6 +2060,10 @@ private IAsyncResult BeginExecuteReaderInternal(CommandBehavior behavior, AsyncC
 
             if (!inRetry)
             {
+                if (asyncWrite)
+                {
+                    _activeConnection.GetOpenTdsConnection().IncrementAsyncCount();
```

Comment:
ON line 2088 RunExecuteReader is called which calls RunExecuteReaderTds which already does something similar to what you are doing here for async cases. Why is that not working? Basically we are ref counting the connection before the execution of the command and that seems to be happening for the code path that you seem to be changing as well. So why do we need this extra increment here? 

```
// If we are in retry, don't increment the Async count. This should have already been set.
            if (async && !inRetry)
            {
                _activeConnection.GetOpenTdsConnection().IncrementAsyncCount();
                decrementAsyncCountOnFailure = true;
            }
```

I think the code is already structured to do what you are doing here, the question is why is the async count getting messed up with the current code.

Reply:
We'll need traces to confirm that for this area, so we can track connection's activity.
I will work on that next week and will update findings here.

Reply:
So I captured traces for the failing scenario, we are setting these counters but something goes terribly wrong when an exception is thrown to user from previous query timeout, take this case for example:

Logs have this:
```
e414c538-cfba-41a2-b26e-b7495e5dc215    82 secs  44   Success
e414c538-cfba-41a2-b26e-b7495e5dc215    82 secs  529  Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
```
When "44" was queried, timeout occurred but User got data as both threads were blocked that handle callbacks "HandleReceiveError" v/s completion callback > and completion callback finished first > that led to Commit being performed, but also Attention signal was sent once "HandleReceiveError" did it's job, about which user is unaware and error stayed in Connection's state when it was pushed to pool.

When "529" was queried, driver queried the database, results came on time, and "async command counter was decremented" as a normal flow, but a pending exception was also found in this connection, due to which the data that was fetched, was not consumed and was pushed back to pool as-is. **This is where the wrong data flow begins.** The async counter seemed to have no relation to whether exception were to be handled before concluding, and by the time exception was handled, it was too late.

![image](https://user-images.githubusercontent.com/13396919/106672638-3af26b00-6565-11eb-80ab-6e6c01b1b8ae.png)

Reply:
So it seems with my changes I'm introducing another counter here on the same method, but it also saved the connection to go back to pool, when "529" as queried.

Reply:
> and error stayed in Connection's state when it was pushed to pool.

I dumped both good and bad objects state from the pool and apart from id's and muxing counters saw no difference so where is the state when it's in the pool?